_posts/2020-04-27-tworzenie-i-usuwanie-indeksow-w-bazie-mongodb-na-produkcji.md

```diff
@@ -0,0 +1,79 @@
+---
+layout:    post
+title:     "Tworzenie i usuwanie indeksów w bazie MongoDB na produkcji"
+date:      2020-04-22 07:00:00 +0100
+published: true
+author:    mgucki
+tags:
+    - tech
+    - mongodb
+    - index
+---
+
+Rankining baz danych pokazują, że już dobrych kilka lat wśród baz typu NoSQL króluje MongoDB i nic nie wskazuje na to, aby miało się to szybko zmienić. Trudno się temu dziwić - MongoDB jest dokumentową bazą danych, ale oferuje dużo więcej poza funkcjonalnością dodawania i pobierania dokumentów.
+
+Mamy do dyspozycji bogaty zestaw narzędzi do zapytań w postaci mechanizmu agregacji. Mamy możliwość nasłuchiwania na zmiany danych w bazie poprzez strumień zmian (_change stream_). Możemy również walidować schemat kolekcji poprzez dostarczony przez nas zestaw reguł. Wersja 4.0 MongoDB dostarcza transakcje obejmujące więcej niż jeden dokument. Już wcześniejszy mechanizm transakcji na poziomie pojedyneczego dokumentu pozwala na tworzenie produkcyjnych aplikacji przy wykorzystaniu funkcji typy znajdź i usuń, podmień oraz zaktualizuj (_findOneAndDelete_, _findOneAndReplace_, _findOneAndUpdate_). W tym przypadku musimy zaprojektować odpowiedni schemat bazy.
+
+Dzięki shardingowi MongoDB pozwala przechowywać i przetwarzać ogromne ilości danych bez znacznego spadku wydajności. Poza tym dzięki redundancji w klastrze (_replica set_) dane są niemalże zawsze dostępne. Co więcej, dokumentacja MongoDB wprost mówi, że klaster z redundacją danych powinien być podstawą każdej instalacji produkcyjnej.
+
+Każdy system jest rozwijany i zmienia się w odpowiedzi na nowe wymagania biznesowe. Z tego powodu może się zdarzyć, że konieczne będą zmiany w zapytaniach do bazy i co za tym idzie również mogą być wymagane zmiany indeksów. Staniemy wtedy przed wyzwaniem zmiany indeksów w systemie produkcyjnym. Może się zdarzyć, że jeden okaże się niepotrzebny i będziemy chcieli go usunąć, aby zwolnić miejsce. Drugi indeks trzeba będzie stworzyć, aby przyspieszyć wykonywanie nowych zapytań. Tworzenie indeksu może trwać bardzo długo. Nie chcemy na ten czas wyłączać produkcji. Poniższy tekst wyjaśnienia jak stworzyć nowy i usunąć stary indeks bez większych zawirowań na produkcji.
+
+Aby cała operacja odbyła się bez problemów, powinniśmy wykonać ją w następującej kolejności:
+- utworzenie indeksu pod nowe zapytania w systemie,
+- aktualizacja części biznesowej systemu,
+- usunięcie starego indeksu, o ile nie jest już używany.
+
+# Tworzenie indeksu w sposób _rolling_
```

Comment:
Tworzenie indeksu w sposób _rolling_ ==> Tworzenie indeksu w Mongo DB w sposób _rolling_
To ze względu na pozycjonowanie (Mongo powinno się pojawić w co najmniej jednym nagłówku, by wyżej pozycjonować się w Google)

Reply:
Poprawione